service/matching/matcher.go

```diff
@@ -658,3 +661,18 @@ func (tm *TaskMatcher) emitForwardedSourceStats(
 func (tm *TaskMatcher) timeSinceLastPoll() time.Duration {
 	return time.Since(time.Unix(0, tm.lastPoller.Load()))
 }
+
+// contextWithCancelOnChannelClose returns a child Context and CancelFunc just like
+// context.WithCancel, but additionally propagates cancellation from another channel (besides
+// the parent's cancellation channel).
+func contextWithCancelOnChannelClose(parent context.Context, closeC <-chan struct{}) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(parent)
+	go func() {
+		select {
+		case <-closeC:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+	return ctx, cancel
+}
```

Comment:
I feel like we could extract some utility here. What we need is a context cancel helper (or some other name) that can wrap a context with cancel, it keep track of all the cancel method in an internal collection. When the cancel is called it remove the cancel from that internal collection. It also expose a CancelAll() method that when called will iterate and call all cancelled that are tracked in its internal collection. Then we just need to create one of such helper for each task queue manager, and when unload task queue manager call that CancelAll().

Reply:
I've had this same thought before. I think we could put it in `common/channel`.

Reply:
What would be the benefit of that approach over this? It saves a goroutine, but at the cost of extra bookkeeping and more code. Also worrying about race conditions (calling WithCancel after CancelAll), where this handles the race automatically.

This way is purely local, and essentially lets the go runtime do the bookkeeping for us. Note that `context.WithCancel` does basically this exact thing internally to propagate cancellation from parent to child.

Where else would we use that utility?

Reply:
I'm with David here. This avoids coordination and let's the runtime deal with it